internal/hcsoci/assigned_devices.go

```diff
@@ -0,0 +1,150 @@
+			}
+			r.resources = append(r.resources, vpci)
+			vmBusInstanceID := coi.HostingSystem.GetAssignedDeviceParentID(vpci.VMBusGUID)
+			log.G(ctx).WithField("vmbus id", vmBusInstanceID).Info("vmbus instance ID")
+
+			vpciVMBusInstanceIDs = append(vpciVMBusInstanceIDs, vmBusInstanceID)
+		}
+	}
+	if len(vpciVMBusInstanceIDs) == 0 {
+		return nil, fmt.Errorf("no assignable devices on the spec %v", coi.Spec.Windows.Devices)
+	}
+	if err := setupDriversWindows(ctx, coi, r); err != nil {
+		return nil, err
+	}
+	deviceUtilPath, err := setupDeviceUtilTools(ctx, coi, r)
+	if err != nil {
+		return nil, err
+	}
+	return getBusAssignedChildrenDeviceLocationPaths(ctx, coi, vpciVMBusInstanceIDs, deviceUtilPath)
+}
+
+// getBusAssignedChildrenDeviceLocationPaths queries the UVM with the device-util tool with the formatted
+// parent bus device for the children devices' location paths from the uvm's view
+// Returns a slice of WindowsDevices created from the resulting children location paths
+func getBusAssignedChildrenDeviceLocationPaths(ctx context.Context, coi *createOptionsInternal, vmBusInstanceIDs []string, deviceUtilPath string) ([]specs.WindowsDevice, error) {
+	p, l, err := createNamedPipeListener()
+	if err != nil {
+		return nil, err
+	}
+	defer l.Close()
+
+	var pipeResults []string
+	errChan := make(chan error)
+	defer close(errChan)
+
+	go readCsPipeOutput(l, errChan, &pipeResults)
+
+	args := createDeviceUtilCommand(deviceUtilPath, vmBusInstanceIDs)
+	req := &shimdiag.ExecProcessRequest{
+		Args:   args,
+		Stdout: p,
+	}
+	exitCode, err := ExecInUvm(ctx, coi.HostingSystem, req)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to find devices with exit code %d", exitCode)
+	}
+
+	// wait to finish parsing stdout results
+	select {
+	case err := <-errChan:
+		if err != nil {
+			return nil, err
+		}
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+
+	// convert stdout results into windows devices
+	results := []specs.WindowsDevice{}
+	for _, value := range pipeResults {
+		specDev := specs.WindowsDevice{
+			ID:     value,
+			IDType: uvm.VPCILocationPathIDType,
+		}
+		results = append(results, specDev)
+	}
+	log.G(ctx).WithField("parsed devices", results).Info("found child assigned devices")
+	return results, nil
+}
+
+func createNamedPipeListener() (string, net.Listener, error) {
+	g, err := guid.NewV4()
+	if err != nil {
+		return "", nil, err
+	}
+	p := `\\.\pipe\` + g.String()
+	l, err := winio.ListenPipe(p, nil)
+	if err != nil {
+		return "", nil, err
+	}
+	return p, l, nil
+}
+
+// readCsPipeOutput is a helper function that connects to a listener and reads
+// the connection's comma separated outut until done. resulting comma separated
+// values are returned in the `result` param. The `done` param is used by this
+// func to indicate completion.
+func readCsPipeOutput(l net.Listener, errChan chan<- error, result *[]string) {
```

Comment:
Does the data from the UVM exec actually get returned in chunks? I.e does all of it not get returned from one read? If this isnt the case couldnt we just do conn.Read and strings.Split(str, ",")?

Reply:
I think it's better to not assume that logic here. I don't know if the results are returned in chunks but if that ever changes in the future, we would see failures here.

Reply:
We'd see failures in regards to the string splitting logic also then

Reply:
I'm just wondering why you went this route

Reply:
Resolving based on offline conversation.